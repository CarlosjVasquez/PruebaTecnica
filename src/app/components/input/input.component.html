<div class="form-group">
  <label
    class="form-label"
    [class.is-disabled]="disabled"
    [for]="id"
    *ngIf="label"
    >{{ label }}</label
  >
  <input
    class="form-input"
    [class.is-invalid]="error"
    [type]="type"
    [id]="id"
    [name]="name"
    [placeholder]="placeholder"
    [disabled]="disabled"
    [required]="required"
    [ngModel]="value"
    (ngModelChange)="updateValue($event)"
    [minlength]="minlength"
    [maxlength]="maxlength"
    [min]="min"
    [max]="max"
    (input)="onHandleInput($event)"
    (change)="onHandleChange($event)"
  />
  <span class="form-error" *ngIf="error">{{ error }}</span>
</div>
