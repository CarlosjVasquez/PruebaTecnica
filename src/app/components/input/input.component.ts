import { CommonModule } from '@angular/common';
import {
  Component,
  EventEmitter,
  forwardRef,
  Input,
  OnChanges,
  Output,
  SimpleChanges,
} from '@angular/core';
import {
  ControlValueAccessor,
  FormsModule,
  NG_VALUE_ACCESSOR,
  ValidationErrors,
} from '@angular/forms';

@Component({
  selector: 'custom-input',
  standalone: true,
  imports: [CommonModule, FormsModule],
  templateUrl: './input.component.html',
  styleUrl: './input.component.scss',
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => InputComponent),
      multi: true,
    },
  ],
})
export class InputComponent implements ControlValueAccessor, OnChanges {
  @Input() label: string = '';
  @Input() type: string = 'text';
  @Input() placeholder: string = '';
  @Input() id!: string;
  @Input() name!: string;
  @Input() disabled: boolean = false;
  @Input() errors!: ValidationErrors | null | undefined;
  @Input() required!: boolean;
  @Input() minlength!: number;
  @Input() maxlength!: number;
  @Input() min!: number | string;
  @Input() max!: number;
  @Input() onlyLetters: boolean = false;
  @Output() onChangeEventEmitter: EventEmitter<string> = new EventEmitter();
  @Output() onInputEventEmitter: EventEmitter<string> = new EventEmitter();
  value: any;
  error: string = '';

  onChange: any = () => {};
  onTouch: any = () => {};

  ngOnChanges(changes: SimpleChanges): void {
    if (changes['errors']?.currentValue) {
      const newErrors = changes['errors'].currentValue;
      if (newErrors['required']) {
        this.error = 'Este campo es requerido';
        return;
      }
      if (newErrors['min']) {
        this.error = `Este campo debe ser mínimo ${newErrors['min']}`;
        return;
      }
      if (newErrors['minlength']) {
        this.error = `Este campo debe tener mínimo ${newErrors['minlength']['requiredLength']} caracteres`;
        return;
      }
      if (newErrors['customError']) {
        this.error = newErrors['customError']['message'];
      }

      return;
    }
    this.error = '';
  }

  writeValue(value: any): void {
    this.value = value;
    this.onChange(value);
    this.onTouch(value);
  }

  registerOnChange(fn: any): void {
    this.onChange = fn;
  }

  registerOnTouched(fn: any): void {
    this.onTouch = fn;
  }

  setDisabledState(isDisabled: boolean): void {
    this.disabled = isDisabled;
  }

  updateValue(event: string) {
    this.onChange(event);
    this.onTouch(event);
  }

  onHandleChange(event: Event) {
    const target = event.target as HTMLInputElement;
    this.onChangeEventEmitter.emit(target.value);
  }

  onHandleInput(event: Event) {
    const input = event.target as HTMLInputElement;
    if (this.onlyLetters)
      input.value = input.value.replace(/[^a-zA-Z0-9.\-\sáéíóúÁÉÍÓÚ]/g, '');
    this.onInputEventEmitter.emit(input.value);
  }
}
