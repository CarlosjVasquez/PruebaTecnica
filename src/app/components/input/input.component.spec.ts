import { ComponentFixture, TestBed } from '@angular/core/testing';
import {
  FormsModule,
  ReactiveFormsModule,
  NG_VALUE_ACCESSOR,
} from '@angular/forms';
import { CommonModule } from '@angular/common';
import { By } from '@angular/platform-browser';
import { InputComponent } from './input.component';
import { SimpleChanges } from '@angular/core';

describe('InputComponent', () => {
  let component: InputComponent;
  let fixture: ComponentFixture<InputComponent>;
  let compiled: HTMLElement;

  interface MockFormControls {
    [key: string]: any;
  }

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      providers: [InputComponent],
      imports: [CommonModule, FormsModule, ReactiveFormsModule],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(InputComponent);
    component = fixture.componentInstance;
    compiled = fixture.nativeElement;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('match to snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should render label when provided', () => {
    const label = 'Test Label';
    component.label = label;
    fixture.detectChanges();
    const labelElement = fixture.debugElement.query(By.css('label'));
    expect(labelElement.nativeElement.textContent).toContain(label);
  });

  test('should not display error message when no errors are present', () => {
    component.errors = null;

    fixture.detectChanges();

    const errorElement = fixture.debugElement.query(By.css('.error-message'));
    expect(errorElement).toBeFalsy();
  });

  test('should update error message when errors change', () => {
    const changes = {
      errors: {
        currentValue: {
          minlength: { requiredLength: 5 },
        },
      },
    } as unknown as SimpleChanges & MockFormControls;

    component.ngOnChanges(changes);
    expect(component.error).toBe('Este campo debe tener mÃ­nimo 5 caracteres');
  });

  test('should update error message for customError', () => {
    const changes = {
      errors: {
        currentValue: {
          customError: { message: 'Custom error message' },
        },
      },
    } as unknown as SimpleChanges & MockFormControls;

    component.ngOnChanges(changes);
    expect(component.error).toBe('Custom error message');
  });

  test('should clear error message when no errors are present', () => {
    component.error = 'Previous error message';
    const changes = {
      errors: {
        currentValue: null,
      },
    } as unknown as SimpleChanges & MockFormControls;

    component.ngOnChanges(changes);
    expect(component.error).toBe('');
  });

  test('should call onChange and onTouch when updateValue is called', () => {
    const onChangeSpy = jest.spyOn(component, 'onChange');
    const onTouchSpy = jest.spyOn(component, 'onTouch');
    const value = 'New Value';

    component.updateValue(value);
    expect(onChangeSpy).toHaveBeenCalledWith(value);
    expect(onTouchSpy).toHaveBeenCalledWith(value);
  });

  test('should emit onChangeEventEmitter on handle change', () => {
    const value = 'New Value';
    const event = { target: { value } } as unknown as Event & MockFormControls;

    jest.spyOn(component.onChangeEventEmitter, 'emit');
    component.onHandleChange(event);
    expect(component.onChangeEventEmitter.emit).toHaveBeenCalledWith(value);
  });

  test('should emit onInputEventEmitter on handle input', () => {
    const value = 'New Input Value';
    const event = { target: { value } } as unknown as Event & MockFormControls;

    jest.spyOn(component.onInputEventEmitter, 'emit');
    component.onHandleInput(event);
    expect(component.onInputEventEmitter.emit).toHaveBeenCalledWith(value);
  });

  test('should replace non-letter characters if onlyLetters is true', () => {
    const inputValue = 'abc123!@#';
    const cleanedValue = 'abc123';
    const event = { target: { value: inputValue } } as unknown as Event &
      MockFormControls;
    const input = event.target as HTMLInputElement;

    component.onlyLetters = true;
    jest.spyOn(component.onInputEventEmitter, 'emit');
    component.onHandleInput(event);
    expect(input.value).toBe(cleanedValue);
    expect(component.onInputEventEmitter.emit).toHaveBeenCalledWith(
      cleanedValue
    );
  });
});
