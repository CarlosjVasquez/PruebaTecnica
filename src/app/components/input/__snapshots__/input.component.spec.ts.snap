// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputComponent match to snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="form-group"
  >
    
    <input
      class="form-input ng-untouched ng-pristine ng-valid"
      id="undefined"
      max="undefined"
      min="undefined"
      placeholder=""
      type="text"
    />
    
  </div>
</div>
`;
